crates/spawner/src/destroyer.rs

```diff
@@ -9,15 +9,51 @@ pub(crate) struct DestroyerPlugin;
 impl Plugin for DestroyerPlugin {
     fn build(&self, app: &mut App) {
         app.add_system(
+            find_dead
+                .in_base_set(GameSet::Update)
+                .run_if(in_state(AppState::InGame))
+                .in_set(SpawnerSet::Destroyer)
+                .in_set(DestroyerSet::FindDead),
+        )
+        .add_system(
             destroy
                 .in_base_set(GameSet::Update)
                 .run_if(in_state(AppState::InGame))
-                .in_set(SpawnerSet::Destroyer),
+                .after(DestroyerSet::SendData)
+                .in_set(SpawnerSet::Destroyer)
+                .in_set(DestroyerSet::DestroyDead),
         );
     }
 }
 
-fn destroy(
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum DestroyerSet {
+    FindDead,
+    SendData,
+    DestroyDead,
+}
+
+/// Special marker trait which when used on a component the despawner sends an event with the data
+pub trait DespawnEventProducer {
```

Comment:
Thinking about it more: we can introduce `DespawnEventsPlugin<T: ReadOnlyWorldQuery>` ([ReadOnlyWorldQuery](https://docs.rs/bevy/latest/bevy/ecs/query/trait.ReadOnlyWorldQuery.html#)) which would automatically send events `DespawnedEvent<<T as WorldQuery>::Item>`. We might need to limit Item to be Copy so we can easily produce the events.

This would make it much more versatile:\

* one would be able to query any combination of components
* the events could be processed during whole frame (even after Update base set when the entity is already despawned)
* there would not be any trait one has to implement, all needed would be to add `DespawnEventsPlugin<T>`